lxd/storage/backend_lxd.go

```diff
@@ -3731,19 +3735,135 @@ func (b *lxdBackend) ImportBucket(projectName string, poolVol *backupConfig.Conf
 	revert.Add(func() { _ = BucketDBDelete(context.TODO(), b, bucketID) })
 
 	// Get the bucket name on storage.
-	storageBucketName := project.StorageVolume(projectName, poolVol.Bucket.Name)
+	storageBucketName := project.StorageVolume(projectName, bucket.Name)
 	storageBucket := b.GetVolume(drivers.VolumeTypeBucket, drivers.ContentTypeFS, storageBucketName, bucketConfig)
 
 	err = b.driver.ValidateVolume(storageBucket, false)
 	if err != nil {
 		return nil, err
 	}
 
+	// Recover bucket keys.
+
+	// Start minio process.
+	minioProc, err := b.ActivateBucket(projectName, bucket.Name, op)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize minio client object.
+	adminClient, err := minioProc.AdminClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, ctxCancel := context.WithTimeout(context.TODO(), time.Duration(time.Second*30))
+	defer ctxCancel()
+
+	// Export IAM data (response is ZIP file).
+	iamReader, err := adminClient.ExportIAM(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer iamReader.Close()
+
+	iamBytes, err := ioutil.ReadAll(iamReader)
+	if err != nil {
+		return nil, err
+	}
+
+	iamZipReader, err := zip.NewReader(bytes.NewReader(iamBytes), int64(len(iamBytes)))
+	if err != nil {
+		return nil, err
+	}
+
+	// We are interesed only in a json file that contains service accounts.
+	// Find that file and extract service accounts.
+	svcAccounts := map[string]madmin.Credentials{}
+	for _, file := range iamZipReader.File {
+		if file.Name != "iam-assets/svcaccts.json" {
+			continue
+		}
+
+		f, err := file.Open()
+		if err != nil {
+			return nil, err
+		}
+
+		defer f.Close()
+
+		fContent, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+
+		err = json.Unmarshal(fContent, &svcAccounts)
+		if err != nil {
+			return nil, err
+		}
+
+		break
+	}
+
+	adminPolicy, err := s3.BucketPolicy(bucket.Name, "admin")
+	if err != nil {
+		return nil, err
+	}
+
+	// Create bucket key DB entry for each service account.
+	for _, creds := range svcAccounts {
+		key := api.StorageBucketKeysPost{
+			Name: creds.AccessKey,
+			StorageBucketKeyPut: api.StorageBucketKeyPut{
+				Description: "Recovered bucket key",
+				Role:        "read-only",
+				AccessKey:   creds.AccessKey,
+				SecretKey:   creds.SecretKey,
+			},
+		}
+
+		svcAccountInfo, err := adminClient.InfoServiceAccount(ctx, creds.AccessKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// If service account policy equals to the admin bucket policy, set bucket key role to admin.
+		if comparePolicy(string(adminPolicy), svcAccountInfo.Policy) {
+			key.Role = "admin"
```

Comment:
What do you think about defining some constant(s) that we can use here and in where we create/manage the key roles so we can easily see in the future what parts of the code base deal with bucket key roles.

Reply:
Agree, will do that.